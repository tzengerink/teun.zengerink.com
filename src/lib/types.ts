export enum Width {
  Mobile = 768,
  Desktop = 1280,
}

export interface ConfigCaption {
  key: string
  caption: string
}

export interface ConfigItem {
  title: string
  statement?: string
  isArchived?: boolean
  captions?: ConfigCaption[]
}

export interface Export {
  width: Width
  url: string
}

export interface Photo {
  key: string
  caption?: string
  exports: Export[]
}

export interface Project {
  title: string
  slug: string
  photos: Photo[]
  statement?: string
  isArchived?: boolean
}

export const MAX_WIDTHS = [Width.Mobile, Width.Desktop]
