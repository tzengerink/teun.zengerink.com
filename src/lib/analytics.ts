export const trackPageView = (): void => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  const method = process.env.NEXT_PUBLIC_PIWIK_APP_ID ? _paq.push : console.info
  method(['trackPageView'])
}
