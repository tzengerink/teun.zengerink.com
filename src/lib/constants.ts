export const TITLE_ROOT = 'Teun Zengerink'
export const TITLE_DIV = '-'
