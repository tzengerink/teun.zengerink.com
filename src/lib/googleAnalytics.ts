declare global {
  interface Window {
    gtag: (...any) => void
  }
}

export const trackPageView = (url: string): void => {
  if (!process.env.NEXT_PUBLIC_TRACKING_ID) return console.info(`[Tracking Pageview] ${url}`)
  window.gtag('config', process.env.NEXT_PUBLIC_TRACKING_ID, {
    page_path: url,
  })
}
