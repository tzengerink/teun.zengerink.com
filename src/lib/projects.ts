import fs from 'fs'
import path from 'path'
import slugify from 'slugify'
import yaml from 'js-yaml'

export enum Width {
  Mobile = 768,
  Desktop = 1280,
}

interface ConfigCaption {
  key: string
  caption: string
}

interface ConfigItem {
  title: string
  statement?: string
  isArchived?: boolean
  captions?: ConfigCaption[]
}

export interface Export {
  width: Width
  url: string
}

export interface Photo {
  key: string
  caption?: string
  exports: Export[]
}
export interface Project {
  title: string
  slug: string
  photos: Photo[]
  statement?: string
  isArchived?: boolean
}

export const MAX_WIDTHS = [Width.Mobile, Width.Desktop]

const generatePhotos = (slug: string, captions: ConfigCaption[]): Photo[] => {
  const photosDirectory = path.join(process.cwd(), 'public/photos', slug, `${Width.Mobile}w`)
  const fileNames = fs.readdirSync(photosDirectory)

  return fileNames.map((filename) => {
    const key = path.basename(filename, '.jpg')
    const caption = captions?.find((c) => c.key == key)

    return {
      key,
      caption: caption?.caption ?? '',
      exports: MAX_WIDTHS.map((width) => ({ width, url: `/photos/${slug}/${width}w/${filename}` })),
    }
  })
}

const generateProjects = (item: ConfigItem[]): Project[] => {
  return item.map((project) => {
    const slug = slugify(project.title, { lower: true })

    return {
      slug,
      title: project.title,
      statement: project.statement ?? '',
      isArchived: project.isArchived ?? false,
      photos: generatePhotos(slug, project.captions),
    }
  })
}

export const getProjects = async (): Promise<Project[]> => {
  let projects: Project[] = []

  try {
    const file = path.join(process.cwd(), 'src', 'config.yml')
    const config = yaml.load(fs.readFileSync(file, 'utf-8'))
    projects = await generateProjects(config)
  } catch (e) {
    console.error(e)
  }

  return projects
}
