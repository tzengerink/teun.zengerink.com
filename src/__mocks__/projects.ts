import { Project } from '../lib/projects'

const projects: Project[] = [
  {
    title: 'My First Project',
    slug: 'my-first-project',
    statement: 'The *statement* for the _project_',
    photos: [
      {
        key: '01',
        exports: [
          { width: 768, url: `/photos/my-first-project/768w/01.jpg` },
          { width: 1280, url: `/photos/my-first-project/1280w/01.jpg` },
        ],
      },
      {
        key: '02',
        exports: [
          { width: 768, url: `/photos/my-first-project/768w/02.jpg` },
          { width: 1280, url: `/photos/my-first-project/1280w/02.jpg` },
        ],
      },
      {
        key: '03',
        exports: [
          { width: 768, url: `/photos/my-first-project/768w/03.jpg` },
          { width: 1280, url: `/photos/my-first-project/1280w/03.jpg` },
        ],
      },
    ],
  },
  {
    title: 'Project without a statement',
    slug: 'no-statement',
    photos: [
      {
        key: '01',
        caption: 'This is _image 1_',
        exports: [
          { width: 768, url: `/photos/no-statement/768w/01.jpg` },
          { width: 1280, url: `/photos/no-statement/1280w/01.jpg` },
        ],
      },
      {
        key: '02',
        caption: 'This is _image 2_',
        exports: [
          { width: 768, url: `/photos/no-statement/768w/02.jpg` },
          { width: 1280, url: `/photos/no-statement/1280w/02.jpg` },
        ],
      },
      {
        key: '03',
        caption: 'This is _image 3_',
        exports: [
          { width: 768, url: `/photos/no-statement/768w/03.jpg` },
          { width: 1280, url: `/photos/no-statement/1280w/03.jpg` },
        ],
      },
      {
        key: '04',
        caption: 'This is _image 4_',
        exports: [
          { width: 768, url: `/photos/no-statement/768w/04.jpg` },
          { width: 1280, url: `/photos/no-statement/1280w/04.jpg` },
        ],
      },
    ],
  },
  {
    title: 'Archived project',
    slug: 'archived-project',
    isArchived: true,
    photos: [
      {
        key: '01',
        exports: [
          { width: 768, url: `/photos/archived-project/768w/01.jpg` },
          { width: 1280, url: `/photos/archived-project/1280w/01.jpg` },
        ],
      },
      {
        key: '02',
        exports: [
          { width: 768, url: `/photos/archived-project/768w/02.jpg` },
          { width: 1280, url: `/photos/archived-project/1280w/02.jpg` },
        ],
      },
    ],
  },
]

export default projects
