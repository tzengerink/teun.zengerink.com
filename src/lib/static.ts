import { createGetProjects, Project } from './projects'

interface Path {
  params: { slug: string[] }
}

export interface Props {
  projects: Project[]
}

const getProjects = createGetProjects()

export const getStaticPaths = async (): Promise<{ paths: Path[]; fallback: boolean }> => {
  const projects = await getProjects()

  return {
    paths: projects.flatMap((project) => {
      const paths = project.photos.map((photo) => ({ params: { slug: [project.slug, photo.key] } }))
      if (project.statement) paths.unshift({ params: { slug: [project.slug] } })
      return paths
    }),
    fallback: false,
  }
}

export const getStaticProps = async (): Promise<{ props: Props }> => ({ props: { projects: await getProjects() } })
