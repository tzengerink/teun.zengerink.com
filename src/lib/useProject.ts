import { Photo, Project } from './projects'
import { useRouter } from 'next/router'

interface UseProject {
  activeKey: string | undefined
  previous: () => void
  next: () => void
}

const NOOP = () => false

export const useProject = (project: Project): UseProject => {
  if (!project?.photos?.length) return { activeKey: undefined, previous: NOOP, next: NOOP }

  const router = useRouter()

  const activeKey = router?.query?.slug?.length > 1 ? router.query?.slug[1] : project.photos[0].key
  const slideIndex = project.photos.findIndex((photo) => photo.key === activeKey)
  const firstPhoto = project.photos[0]
  const lastPhoto = [...project.photos].pop()
  const nextPhoto = project.photos[slideIndex + 1]
  const previousPhoto = project.photos[slideIndex - 1]
  const isLast = activeKey === lastPhoto.key
  const isFirst = activeKey === firstPhoto.key

  const go = (photo: Photo) => router.push(`${`/work/${project.slug}`}/${photo.key}`)
  const previous = () => (isFirst ? go(lastPhoto) : go(previousPhoto))
  const next = () => (isLast ? go(firstPhoto) : go(nextPhoto))

  return { activeKey, previous, next }
}
