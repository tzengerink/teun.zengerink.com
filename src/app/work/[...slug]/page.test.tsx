import { render } from '@testing-library/react'

import Page, { generateMetadata } from './page'
import { getProjects } from '../../../lib/projects'
import { TITLE_DIV, TITLE_ROOT } from '../../../lib/constants'

jest.mock('next/navigation', () => ({ useRouter: () => ({ push: jest.fn() }) }))

describe('generateMetadata', () => {
  it('generates a title and description', async () => {
    const projects = await getProjects()
    const metadata = await generateMetadata({ params: Promise.resolve({ slug: [projects[0].slug, '01'] }) })
    const title = `${TITLE_ROOT} ${TITLE_DIV} ${projects[0].title}`
    expect(metadata).toEqual({ title, description: title })
  })
})

describe('Page', () => {
  it('renders a project slideshow', async () => {
    const projects = await getProjects()
    const { container } = render(await Page({ params: Promise.resolve({ slug: [projects[0].slug, '01'] }) }))
    expect(container).toMatchSnapshot()
  })

  it('renders a page not found message', async () => {
    const { container } = render(await Page({ params: Promise.resolve({ slug: ['non-existent-project', '01'] }) }))
    expect(container).toMatchSnapshot()
  })
})
