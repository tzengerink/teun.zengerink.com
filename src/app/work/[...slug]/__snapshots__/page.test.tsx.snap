// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page renders a page not found message 1`] = `
<div>
  <div
    class="md:absolute md:inset-20 text-center"
  >
    <h2>
      Page not found
    </h2>
    <a
      class="underline"
      href="/"
    >
      Go to homepage
    </a>
  </div>
</div>
`;

exports[`Page renders a project slideshow 1`] = `
<div>
  <div
    class="relative h-[calc(100vh-150px)] md:h-[100vh]"
  >
    <div
      class="absolute top-1/2 translate-y-[-50%] sm:left-1/2 sm:translate-x-[-50%]"
    >
      <div
        class="absolute top-0 bottom-0 left-0 right-1/2 cursor-w-resize"
        data-testid="left-side"
      />
      <div
        class="absolute top-0 bottom-0 right-0 left-1/2 cursor-e-resize"
        data-testid="right-side"
      />
      <img
        alt="Order/Disorder - 01"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center block"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/01.png"
        srcset="/photos/orderdisorder/768w/01.png 768w, /photos/orderdisorder/1280w/01.png 1280w"
      />
      <img
        alt="Order/Disorder - 02"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/02.png"
        srcset="/photos/orderdisorder/768w/02.png 768w, /photos/orderdisorder/1280w/02.png 1280w"
      />
      <img
        alt="Order/Disorder - 03"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/03.png"
        srcset="/photos/orderdisorder/768w/03.png 768w, /photos/orderdisorder/1280w/03.png 1280w"
      />
      <img
        alt="Order/Disorder - 04"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/04.png"
        srcset="/photos/orderdisorder/768w/04.png 768w, /photos/orderdisorder/1280w/04.png 1280w"
      />
      <img
        alt="Order/Disorder - 05"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/05.png"
        srcset="/photos/orderdisorder/768w/05.png 768w, /photos/orderdisorder/1280w/05.png 1280w"
      />
      <img
        alt="Order/Disorder - 06"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/06.png"
        srcset="/photos/orderdisorder/768w/06.png 768w, /photos/orderdisorder/1280w/06.png 1280w"
      />
      <img
        alt="Order/Disorder - 07"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/07.png"
        srcset="/photos/orderdisorder/768w/07.png 768w, /photos/orderdisorder/1280w/07.png 1280w"
      />
      <img
        alt="Order/Disorder - 08"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/08.png"
        srcset="/photos/orderdisorder/768w/08.png 768w, /photos/orderdisorder/1280w/08.png 1280w"
      />
      <img
        alt="Order/Disorder - 09"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/09.png"
        srcset="/photos/orderdisorder/768w/09.png 768w, /photos/orderdisorder/1280w/09.png 1280w"
      />
      <img
        alt="Order/Disorder - 10"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/10.png"
        srcset="/photos/orderdisorder/768w/10.png 768w, /photos/orderdisorder/1280w/10.png 1280w"
      />
      <img
        alt="Order/Disorder - 11"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/11.png"
        srcset="/photos/orderdisorder/768w/11.png 768w, /photos/orderdisorder/1280w/11.png 1280w"
      />
      <img
        alt="Order/Disorder - 12"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/12.png"
        srcset="/photos/orderdisorder/768w/12.png 768w, /photos/orderdisorder/1280w/12.png 1280w"
      />
      <img
        alt="Order/Disorder - 13"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/13.png"
        srcset="/photos/orderdisorder/768w/13.png 768w, /photos/orderdisorder/1280w/13.png 1280w"
      />
      <img
        alt="Order/Disorder - 14"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/14.png"
        srcset="/photos/orderdisorder/768w/14.png 768w, /photos/orderdisorder/1280w/14.png 1280w"
      />
      <img
        alt="Order/Disorder - 15"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/15.png"
        srcset="/photos/orderdisorder/768w/15.png 768w, /photos/orderdisorder/1280w/15.png 1280w"
      />
      <img
        alt="Order/Disorder - 16"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/16.png"
        srcset="/photos/orderdisorder/768w/16.png 768w, /photos/orderdisorder/1280w/16.png 1280w"
      />
      <img
        alt="Order/Disorder - 17"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/17.png"
        srcset="/photos/orderdisorder/768w/17.png 768w, /photos/orderdisorder/1280w/17.png 1280w"
      />
      <img
        alt="Order/Disorder - 18"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/18.png"
        srcset="/photos/orderdisorder/768w/18.png 768w, /photos/orderdisorder/1280w/18.png 1280w"
      />
      <img
        alt="Order/Disorder - 19"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/19.png"
        srcset="/photos/orderdisorder/768w/19.png 768w, /photos/orderdisorder/1280w/19.png 1280w"
      />
      <img
        alt="Order/Disorder - 20"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/20.png"
        srcset="/photos/orderdisorder/768w/20.png 768w, /photos/orderdisorder/1280w/20.png 1280w"
      />
      <img
        alt="Order/Disorder - 21"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/orderdisorder/1280w/21.png"
        srcset="/photos/orderdisorder/768w/21.png 768w, /photos/orderdisorder/1280w/21.png 1280w"
      />
    </div>
  </div>
</div>
`;
